### PR TITLE
Correctly track dirty state of encrypted password

### LINKED ISSUE #664: Right way to add additional password validations?
Perhaps I want to require passwords to be 10 characters long (default out of the box clearance will allow 1 character password). 

I try this in my `User` model:

      validates :password, length: { in: 10..128 }, unless: :skip_password_validation?

If I use `user.update_password`, then it indeed validates using that, returns false with validation errors in `user.errors`, great. 

But the [default Clearance new user registration routines](https://github.com/thoughtbot/clearance/blob/master/app/controllers/clearance/users_controller.rb#L49-L54) do not seem to result in `skip_password?` being unset, so my validation is never called on the password entered by a newly registered user, and they can use it no problem.

But if I skip the `unless: :skip_password_validation?`, of course I'll never be able to save a user unless I'm resetting their password, since `password` is not generally exposed. Which is why `skip_password?` is there in the first place for the [one built-in presence validation](https://github.com/thoughtbot/clearance/blob/master/lib/clearance/user.rb#L150), I gather... except even this one built-in validation is probably not called for newly registering users either?

Is there anything I'm missing? Can you give advice on the best way to add additional password requirements in a way that works well with Clearance instead of fighting it? Thanks!

### PR DESCRIPTION
**Pull request by derekprior**
Password validations only run when `skip_password_validation?` returns false. The idea of that method is that it returns false (allowing conditional password validation) whenever the user requires a password and a change to their password was attempted. Previously, we relied on the developer explicitly having set `password_changing = true` on the instance to detect an attempt to change the password.

This worked fine when the developer used `update_password` to change an
existing users password. However, using other methods to update a
password or even creating a new user (like in the `users#create` action)
did not set this attribute.

This was okay for our built-in validations (just presence) because of
the way `password=` was implemented. However, users that added their own
validations (like validating password length) and properly added the
`unless: :skip_password_validation?` modifier to their validation found
that their validations were not running for new users.

To fix this, we rely on ActiveRecord/ActiveModel's built-in dirty
tracking to check if encrypted password has changed rather than using a
manual attribute. We further tell the user model that calling
`password=` at all will automatically mark the encrypted_password as
dirty, regardless of what the password strategy ultimately does.

Making the change in `User` and calling `super` means that all password
strategies -- even custom strategies -- get this improved behavior.

Now that we're using dirty tracking properly, we can deprecate the
`password_changing` attribute, as it is no longer used.

Closes #664

### CODE REVIEW

**Comment by jrochkind** [comment]
Would it make sense to add controller specs making sure the built-in controllers are succesfully enforcing the built-in presence-validator, as i believe they were not before this change?

**Comment by jrochkind** [comment]
Will there be any backwards compat problems with people perhaps using Rails mass assignment to set 'password' to the empty string, trusting that it will be ignored entirely?  Say an HTML password input field in a form, that resultantly gets ignored if you leave it blank.

**Comment by derekprior** [comment]
> Would it make sense to add controller specs making sure the built-in controllers are succesfully enforcing the built-in presence-validator, as i believe they were not before this change?

I will investigate. I think presence validation was working properly, but not in a directly obvious way. But I will verify this.

> Will there be any backwards compat problems with people perhaps using Rails mass assignment to set 'password' to the empty string, trusting that it will be ignored entirely? Say an HTML password input field in a form, that resultantly gets ignored if you leave it blank.

Due to the way `password=` works, I don't think this would be a problem. Even so, I'd consider this a bug we are fixing.

**Comment by mike-burns** [comment]
I came with the backward compat question, and your answer has convinced me. Nice fix; LGTM.

File: lib/clearance/user.rb [hunk 1]
```diff
@@ -103,12 +100,34 @@ module User
...
+        super
+      end
+
+      def password_changing
+        warn "#{Kernel.caller.first}: [DEPRECATION] " \
+          "The `password_changing` attribute is deprecated. Clearance uses " \
+          "the dirty state of the `encrypted_password` field to track this " \
+          "automatically."
+
+        @password_changing
```
Comment (teoljungberg) [inline lib/clearance/user.rb]: Could this be replaced by a call to `super` instead?

Comment (derekprior) [inline lib/clearance/user.rb]: I'd have to keep the `attr_accessor` in that case, and I find declaring the accessor that I'm going to immediately override to be more confusing than inlining the equivalent functionality.

Comment (teoljungberg) [inline lib/clearance/user.rb]: Right, good point.

**Comment by molfar** [comment]
When I submit form with blank password field, Dirty tracking fails and shows wrong flag that encrypted_password changed. It causes validations problems. Example output:

`{"encrypted_password"=> ["$2a$12$vvCUISnSqEiSy9XfEDCw5.LezVbzGELcRoWTHs79ySW4ERCkalfJy", "$2a$12$vvCUISnSqEiSy9XfEDCw5.LezVbzGELcRoWTHs79ySW4ERCkalfJy"] }`

**Pull request merged**